### ISSUE TITLE
The "forcepull" extended test failure

### ISSUE THREAD

**Issue by mfojtik** [openshift/origin#6299]
Just saw this failure in Jenkins:

https://ci.openshift.redhat.com/jenkins/job/origin_extended/409/testReport/junit/(root)/Extended%20Core/builds__forcepull__ForcePull_from_OpenShift_induced_builds__vs__s2i_command__ForcePull_test_context___ForcePull_test_case_execution/

**Comment by gabemontero** [comment]
OK .. I could not find any additional logs from the ci.openshift, but I believe I was able to reproduce this locally.  The additional stdout should several instances of this occurring prior to the same looking exit statement:

Dec 14 13:42:52.642: INFO: Running 'oc create --namespace=extended-test-forcepull-dnl6s --config=/tmp/openshift-extended-tests/core/extended-test-forcepull-dnl6s-user.kubeconfig -f /home/gmontero/go/src/github.com/openshift/origin/test/extended/fixtures/forcepull-setup.json'
Dec 14 13:42:52.687: INFO: Error running &{/home/gmontero/go/src/github.com/openshift/origin/_output/local/bin/linux/amd64/oc [oc create --namespace=extended-test-forcepull-dnl6s --config=/tmp/openshift-extended-tests/core/extended-test-forcepull-dnl6s-user.kubeconfig -f /home/gmontero/go/src/github.com/openshift/origin/test/extended/fixtures/forcepull-setup.json] []   imagestream "forcepull-extended-test-builder" created
The BuildConfig "forcepull-bldr" is invalid.
spec.strategy: invalid value '<*>(0xc20b4431a0){DockerStrategy:<*>(0xc2103c57c0){From:<nil> PullSecret:<nil> NoCache:false Env:<nil> ForcePull:false} SourceStrategy:<*>(0xc20cb468f0){From:{Kind:ImageStreamTag Namespace: Name: UID: APIVersion: ResourceVersion: FieldPath:} PullSecret:<nil> Env:<nil> Scripts: Incremental:false ForcePull:false} CustomStrategy:<nil>}', Details: must provide a value for exactly one of sourceStrategy, customStrategy, or dockerStrategy

The image hierarchy here ultimately depends on the centos7 image, which the team has uncovered is broken, as seen by the recent overall failures on merge_pull_requests_origin.

I'll leave this open for now and retry this after the centos7 situation gets sorted out, but I suspect we won't hit this when centos7 is fixed.

Of course let me know if I am way off base here.

@bparees - fyi

**Comment by bparees** [comment]
definitely not related to centos.

``` Details: must provide a value for exactly one of sourceStrategy, customStrategy, or dockerStrategy```

is the crux of this issue, probably introduced by my changes to the the validation logic when i removed the "Type" field from the Strategy object.

the test's forcepull-setup.json is defining a BuildConfig w/ a Strategy of type "Docker" but does not provide a Docker strategy, instead it provides a SourceStrategy:
```
		"strategy": {
		    "type": "Docker",
		    "sourceStrategy": {
			"nocache": true
		    }
		},
```
not clear to me why it's doing that (or how it was working before, since "nocache" is part of the DockerStrategy, not the SourceStrategy, so if that's necessary, it was having no effect)

So basically at this point two things need to happen:

1) the buildConfig definition needs to be cleaned up
2) it'd be good to understand why validation is reporting the error it is, because that's not the error i'd expect it to report.  In fact i would have expected no error because conversion would just strip the Type, and then you'd be left with a BuildConfig that defines a SourceStrategy and nothing else, which is valid (except that your SourceStrategy isn't fully defined, so you'd get errors for that).

i will look into (2), you can fix (1).

**Comment by bparees** [comment]
ok well the reason you're getting that particular validation error is because of:
https://github.com/openshift/origin/blob/master/pkg/build/api/v1/conversion.go#L181-L188

which initializes the DockerStrategy object by default when the Type is Docker.  That, plus the SourceStrategy object you're already providing results in 2 strategies being provided, which is what the error tells you.

unfortunately we can't get rid of the defaulting, because as of v1 api it's valid to just specify a Type: Docker with no actual dockerstrategy object (there are no required fields so you don't technically need to provide one... we get your Dockerfile from the source/git input and that's all we need).

So i don't think there's anything to change for (2), you just need to fix your BuildConfig.

**Comment by gabemontero** [comment]
ok - thx for the insight - I had interpreted those hex numbers as image id's and went to the centos issue, but in hindsight with your explanation, they must be pointers; 

the json went through a couple of iterations / evolutions around conversion and pruning from the sample app it was originally based off of - some bits in the various conversions must have got turned around and the prior validations did not catch the mix up as you noted.

Will get a pull going shortly.
